### PR TITLE
Only calculate hash once for Marker objects

### PR DESCRIPTION
**Pull request by bdraco**
before
validate_mapping_before

after
validate_mapping_after

### CODE REVIEW

File: voluptuous/__init__.py [hunk 1]
```diff
@@ -8,5 +8,5 @@
 
 # fmt: on
 
-__version__ = '0.14.2'
+__version__ = '0.13.1'
```
Comment (bdraco) [inline voluptuous/__init__.py:11]:
```suggestion
__version__ = '0.14.2'
```

File: voluptuous/schema_builder.py [hunk 2]
```diff
@@ -1244,16 +1244,22 @@ class Remove(Marker):
...
     """
 
+    def __init__(
+        self,
+        schema_: Schemable,
+        msg: typing.Optional[str] = None,
+        description: typing.Optional[str] = None,
+    ) -> None:
+        super().__init__(schema_, msg, description)
+        self.__hash__ = lru_cache(maxsize=None)(lambda: object.__hash__(self))
```
Comment (alecthomas) [inline voluptuous/schema_builder.py:1254]: Why is this necessary? Doesn't `Marker.__hash__` cover this?

**Review by alecthomas — APPROVED** [review]

File: voluptuous/schema_builder.py (hunk 2 above)
Comment (bdraco) [reply voluptuous/schema_builder.py:1254]: Its overriding the hash here to use the base `object.__hash__` instead of `Marker.__hash__`

File: voluptuous/schema_builder.py [hunk 3]
```diff
@@ -1036,6 +1038,7 @@ def __init__(
         self._schema = Schema(schema_)
         self.msg = msg
         self.description = description
+        self.__hash__ = lru_cache(maxsize=None)(lambda: hash(schema_))
```
Comment (alecthomas) [inline voluptuous/schema_builder.py:1041]: It seems like it would be lower overhead to use `functools.cache` here?

**Comment by bdraco** [comment]
I'm letting this run over the weekend on my production Home Assistant so I can examine the cache on Monday once its run for a bit.  If all still looks well, I'll mark it ready for review than

File: voluptuous/schema_builder.py (hunk 3 above)
Comment (bdraco) [reply voluptuous/schema_builder.py:1041]: Yes it would but it was only added in 3.9 and the tests are still running on 3.8

Comment (bdraco) [reply voluptuous/schema_builder.py:1041]: Happy to do a PR to remove 3.8 support since it goes EOL in ~4 months anyways

Comment (alecthomas) [reply voluptuous/schema_builder.py:1041]: Let's do it, I'm okay with that.

Comment (bdraco) [reply voluptuous/schema_builder.py:1041]: Actually I'm not sure it matters much since `cache` does the same thing under the hood and the native 
https://github.com/python/cpython/blob/1e7903d549716eb8034167302d4a718e42460b4b/Lib/functools.py#L653

It all end up back in native code either way https://github.com/python/cpython/blob/1e7903d549716eb8034167302d4a718e42460b4b/Lib/functools.py#L642C28-L642C46

Comment (bdraco) [reply voluptuous/schema_builder.py:1041]: Still cleaner to use `@cache` though

Comment (bdraco) [reply voluptuous/schema_builder.py]: 65ec763d67a18e3c43d7294f4ce6d902105da984

**Comment by ds-cbo** [comment]
@bdraco Out of curiosity a side-question: What did you use to make those graphs?

**Comment by bdraco** [comment]
They are made using cProfile and than it's converted to a callgrind file and opened in qcachegrind 

More details at https://www.home-assistant.io/integrations/profiler/

**Comment by bdraco** [comment]
Thanks. Looking forward to this one in a production HA
